### DISCUSSION TITLE
How do modules work?

### DISCUSSION THREAD

**Discussion by roopeshsn** [deepnoodle-ai/risor#191] [Q&A] [answered]
I have gone through some modules (k8s, uuid, etc.). The common thing I observed is that functions were written that act as a bridge between the original go package and risor, which return a type of object (NewInt, FromGoType, etc.). It would be better If I got to know how these modules work or some hints so that I continue to see how things work.

**Comment by myzie** [comment, marked as answer]
Hi @roopeshsn, you're definitely on the right track.

Luckily, there's really nothing magical about modules. As you've seen, the `object.Object` interface type in Risor is key. Risor's built-in types implement Object and similarly any new Go types you introduce that implement Object can easily be used in Risor.

The Module type is implemented here:
https://github.com/risor-io/risor/blob/main/object/module.go

It's a simple implementation, with the `GetAttr` method looking up fields contained by the module. 

Functions in Risor implement Object as well, via the [Builtin](https://github.com/risor-io/risor/blob/main/object/builtin.go) type and the [Function](https://github.com/risor-io/risor/blob/main/object/function.go) type. The former is for wrapping Go functions, the latter is for compiled Risor functions.

Commonly a Module object then will contain multiple Builtin functions, as is the case with the `math` and `strings` modules for example. And this is the typical way the Module objects are initialized:

```go
func Module() *object.Module {
	return object.NewBuiltinsModule("math", map[string]object.Object{
		"abs":    object.NewBuiltin("abs", Abs),
        }
}
```
